### PR TITLE
[FLINK-21655] [table-planner-blink] Fix incorrect simplification for coalesce call on a groupingsets' result

### PR DESCRIPTION
**Pull request by lincoln-lil**
## What is the purpose of the change

Fix incorrect simplification for coalesce call on a groupingsets' result

## Brief change log

Fix nullability inference logic in ExpandUtil

## Verifying this change
Added ut & it  cases

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no )
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no )
  - The serializers: ( no )
  - The runtime per-record code paths (performance sensitive): ( no )
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: ( no )
  - The S3 file system connector: ( no )

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by lincoln-lil** [comment]
The test failure is irrelevant module kafka_gelly testing timeout.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/calcite/Expand.scala [hunk 1]
```diff
@@ -79,7 +83,27 @@ abstract class Expand(
...
+        project.zipWithIndex.map {
+          f => {
+            val nullable = fieldNullableMap.get(f._2).getOrElse(f._1.getType.isNullable)
+            fieldNullableMap.put(f._2, nullable || f._1.getType.isNullable)
+          }
+        }
+    }
+    val typeList = ListBuffer[RelDataType]()
+    val typeFactory = cluster.getTypeFactory
+    projects.get(0).zipWithIndex.map {
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/calcite/Expand.scala:99]: another case we should consider is Expand can be more generic (not only used for distinct agg, but also for other case), after that the fields in the Expand node may be different type, so we should check whether a column's types in all expanded rows are the same or we should derive the common type for each column.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/ExpandUtil.scala [hunk 2]
```diff
@@ -19,16 +19,11 @@
 package org.apache.flink.table.planner.plan.utils
 
 import org.apache.flink.table.planner.calcite.FlinkRelBuilder
-import org.apache.flink.table.planner.plan.nodes.calcite.{Expand, LogicalExpand}
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/ExpandUtil.scala:22]: nit: reformat imports order

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/ExpandUtil.scala [hunk 3]
```diff
@@ -190,15 +172,12 @@ object ExpandUtil {
   def createExpandProjects(
       rexBuilder: RexBuilder,
       inputType: RelDataType,
-      outputType: RelDataType,
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/utils/ExpandUtil.scala:193]: the java doc should also be updated

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdDistinctRowCount.scala [hunk 4]
```diff
@@ -223,7 +223,7 @@ class FlinkRelMdDistinctRowCount private extends MetadataHandler[BuiltInMetadata
         val groupKeyOfCurrentProject = new JArrayList[Int]()
         groupKeySkipExpandId.foreach { key =>
           project.get(key) match {
-            case literal: RexLiteral if literal.isNull => // do nothing
+            case literal: RexLiteral if literal.isNull || literal.isAlwaysFalse => // do nothing
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdDistinctRowCount.scala:226]: there are some corner case ?

**Review by godfreyhe — COMMENTED** [review]
Thanks for the contribution @lincoln-lil , I left some comments

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/calcite/Expand.scala [hunk 5]
```diff
@@ -59,10 +60,13 @@ abstract class Expand(
     if (projects.size() <= 1) {
       return litmus.fail("Expand should output more than one rows, otherwise use Project.")
     }
-    if (projects.exists(_.size != outputRowType.getFieldCount)) {
+    if (null == outputFieldNames) {
+      return litmus.fail("Expand should have none empty output field names.")
+    }
```
[the remark this answers is not in the document]
Comment (lincoln-lil) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/calcite/Expand.scala:65]: make sense

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/calcite/Expand.scala (hunk 1 above)
Comment (lincoln-lil) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/calcite/Expand.scala:99]: good point！I'll try to investigate the two ways.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdDistinctRowCount.scala (hunk 4 above)
Comment (lincoln-lil) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdDistinctRowCount.scala:226]: It's unnecessary now, just a patch for the boolean type literal which introduced from the previous offline discussion.
I'll remove it.

**Review by godfreyhe — APPROVED** [review]

**Pull request closed without merging**